src/com/ms/silverking/cloud/dht/client/impl/AsyncOperationImpl.java

```diff
@@ -27,446 +28,490 @@
-            lock.unlock();
-        }
-        return complete;
+    if (notifyListeners) {
+      // Using calling thread to trigger updates for completion, but only for completion
+      if (opState != OperationState.INCOMPLETE) {
+        notifyListeners(_listeners);
+      } else {
+        notificationWorker.update(this, ImmutableSet.copyOf(_listeners));
+      }
     }
-    
-    public boolean poll(long duration, TimeUnit timeUnit) {
-        boolean    complete;
-        
-        lock.lock();
+  }
+
+  void _waitForCompletion() throws OperationException {
+    lock.lock();
+    try {
+      while (!result.isComplete()) {
         try {
-            while (!result.isComplete()) {
-                try {
-                    if (!spin) {
-                        cv.await(duration, timeUnit);
-                    } else {
-                        cv.awaitNanos(spinDurationNanos);
-                    }
-                } catch (InterruptedException ie) {
-                }
-            }
-            complete = result.isComplete();
-        } finally {
-            lock.unlock();
+          Log.fine("activeOp awaiting ", this);
+          if (!spin) {
+            cv.await();
+          } else {
+            cv.awaitNanos(spinDurationNanos);
+          }
+        } catch (InterruptedException ie) {
         }
-        return complete;
-    }
-    
-    protected void setSent() {
-        sent = true;
+      }
+      if (isFailure(result)) {
+        throwFailedException();
+      }
+    } finally {
+      lock.unlock();
     }
-    
-    protected boolean getSent() {
-        return sent;
-    }
-    
-    abstract void addToEstimate(MessageEstimate estimate);
-    abstract MessageEstimate createMessageEstimate();
-    abstract ProtoMessageGroup createProtoMG(MessageEstimate estimate);
-    abstract ProtoMessageGroup createMessagesForIncomplete(ProtoMessageGroup protoMG, List<MessageGroup> messageGroups, MessageEstimate estimate);
-    
-    @Override
-    public String toString() {
-        StringBuilder    sb;
-        
-        sb = new StringBuilder();
-        sb.append(super.toString());
-        sb.append(operation.getOpType());
-        sb.append(':');
-        sb.append(getUUID());
-        sb.append(':');
-        sb.append(result);
-        return sb.toString();
+  }
+
+  public boolean poll() {
+    boolean complete;
+
+    lock.lock();
+    try {
+      complete = result.isComplete();
+    } finally {
+      lock.unlock();
     }
-    
-    // for debugging
-    protected final String objectToString() {
-        return super.toString();
+    return complete;
+  }
+
+  public boolean poll(long duration, TimeUnit timeUnit) {
+    boolean complete;
+
+    lock.lock();
+    try {
+      while (!result.isComplete()) {
+        try {
+          if (!spin) {
+            cv.await(duration, timeUnit);
+          } else {
+            cv.awaitNanos(spinDurationNanos);
+          }
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)